File: annotation-processor/src/test/java/me/bristermitten/mittenlib/annotations/util/TypeSpecUtilTest.java
package me.bristermitten.mittenlib.annotations.util;

import static org.junit.jupiter.api.Assertions.*;

import com.palantir.javapoet.MethodSpec;
import java.lang.annotation.Repeatable;
import java.lang.annotation.Retention;
import java.lang.annotation.RetentionPolicy;
import org.junit.jupiter.api.Test;

class TypeSpecUtilTest {

    @Retention(RetentionPolicy.RUNTIME)
    @interface NonRepeatableAnnotation {}

    @Retention(RetentionPolicy.RUNTIME)
    @Repeatable(RepeatableAnnotations.class)
    @interface RepeatableAnnotation {}

    @Retention(RetentionPolicy.RUNTIME)
    @interface RepeatableAnnotations {
        RepeatableAnnotation[] value();
    }

    @Test
    void testMethodAddAnnotationAddsNonRepeatableOnce() {
        MethodSpec.Builder builder = MethodSpec.methodBuilder("test");
        TypeSpecUtil.methodAddAnnotation(builder, NonRepeatableAnnotation.class);
        assertEquals(1, builder.build().annotations().size());

        TypeSpecUtil.methodAddAnnotation(builder, NonRepeatableAnnotation.class);
        assertEquals(1, builder.build().annotations().size()); // Should not add again
    }

    @Test
    void testMethodAddAnnotationAddsRepeatableMultipleTimes() {
        MethodSpec.Builder builder = MethodSpec.methodBuilder("test");
        TypeSpecUtil.methodAddAnnotation(builder, RepeatableAnnotation.class);
        assertEquals(1, builder.build().annotations().size());

        TypeSpecUtil.methodAddAnnotation(builder, RepeatableAnnotation.class);
        assertEquals(2, builder.build().annotations().size()); // Should add again
    }
}

File: annotation-processor/src/test/scala/me/bristermitten/mittenlib/annotations/util/TypeSpecUtilTest.scala
package me.bristermitten.mittenlib.annotations.util

import com.palantir.javapoet.MethodSpec
import org.scalatest.funsuite.AnyFunSuite
import org.scalatest.matchers.should.Matchers

class TypeSpecUtilTest extends AnyFunSuite with Matchers {

  test("testMethodAddAnnotationAddsNonRepeatableOnce") {
    val builder = MethodSpec.methodBuilder("test")
    TypeSpecUtil.methodAddAnnotation(
      builder,
      classOf[IntegrationTests.NonRepeatableAnnotation]
    )
    builder.build().annotations().size() shouldBe 1

    TypeSpecUtil.methodAddAnnotation(
      builder,
      classOf[IntegrationTests.NonRepeatableAnnotation]
    )
    builder.build().annotations().size() shouldBe 1 // Should not add again
  }

  test("testMethodAddAnnotationAddsRepeatableMultipleTimes") {
    val builder = MethodSpec.methodBuilder("test")
    TypeSpecUtil.methodAddAnnotation(
      builder,
      classOf[IntegrationTests.RepeatableAnnotation]
    )
    builder.build().annotations().size() shouldBe 1

    TypeSpecUtil.methodAddAnnotation(
      builder,
      classOf[IntegrationTests.RepeatableAnnotation]
    )
    builder.build().annotations().size() shouldBe 2 // Should add again
  }
}
